pkg/engine/lifecycletest/transformation_test.go

```diff
@@ -968,3 +969,74 @@ func TestTransformInvokeTransformProvider(t *testing.T) {
 	assert.NotNil(t, snap)
 	assert.Equal(t, 1, len(snap.Resources)) // expect no default provider to be created for the invoke
 }
+
+// Regression test for https://github.com/pulumi/pulumi/issues/19904. Registering a transform that depends on a resource
+// that is registered after the transform starts running should not hang.
+func TestTransformOrdering(t *testing.T) {
+	t.Parallel()
+
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{}, nil
+		}),
+	}
+
+	var implicitProvider string
+	programF := deploytest.NewLanguageRuntimeF(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		pcs := &promise.CompletionSource[*deploytest.RegisterResourceResponse]{}
+
+		callbacks, err := deploytest.NewCallbacksServer()
+		require.NoError(t, err)
+		callback, err := callbacks.Allocate(
+			TransformFunction(func(name, typ string, custom bool, parent string,
+				props resource.PropertyMap, opts *pulumirpc.TransformResourceOptions,
+			) (resource.PropertyMap, *pulumirpc.TransformResourceOptions, error) {
+				if name == "resA" && opts.Provider == "" {
+					provider, err := pcs.Promise().Result(context.Background())
+					if err != nil {
+						return nil, nil, err
+					}
+					implicitProvider = string(provider.URN) + "::" + provider.ID.String()
+					opts.Provider = implicitProvider
+				}
+
+				return props, opts, nil
+			}))
+		require.NoError(t, err)
+
+		err = monitor.RegisterStackTransform(callback)
+		require.NoError(t, err)
+
+		resA := promise.Run(func() (*deploytest.RegisterResourceResponse, error) {
+			return monitor.RegisterResource("pkgA:m:typA", "resA", true)
+		})
+
+		go func() {
+			resp, err := monitor.RegisterResource("pulumi:providers:pkgA", "implicit", true)
+			if err != nil {
+				pcs.MustReject(err)
+				return
+			}
+			pcs.MustFulfill(resp)
+		}()
+
+		_, err = resA.Result(context.Background())
+		require.NoError(t, err)
+
+		return nil
+	})
+
+	hostF := deploytest.NewPluginHostF(nil, nil, programF, loaders...)
+	p := &lt.TestPlan{
+		Options: lt.TestUpdateOptions{T: t, HostF: hostF},
+		Steps: []lt.TestStep{
+			{
+				Op: Update,
+			},
+		},
+	}
+	snap := p.Run(t, nil)
+	assert.NotNil(t, snap)
+	assert.Equal(t, 2, len(snap.Resources))
```

Comment:
```suggestion
	require.NotNil(t, snap)
	require.Equal(t, 2, len(snap.Resources))
```

Both of the above need to be require to avoid panics on test failures